grakn-graql/src/main/java/ai/grakn/graql/internal/pattern/property/HasResourceTypeProperty.java

```diff
@@ -63,93 +64,87 @@
  *
  * @author Felix Chapman
  */
-public class HasResourceTypeProperty extends AbstractVarProperty implements NamedProperty {
+@AutoValue
+public abstract class HasResourceTypeProperty extends AbstractVarProperty implements NamedProperty {
 
-    private final VarPatternAdmin resourceType;
+    abstract VarPatternAdmin resourceType();
 
-    private final VarPatternAdmin ownerRole;
-    private final VarPatternAdmin valueRole;
-    private final VarPatternAdmin relationOwner;
-    private final VarPatternAdmin relationValue;
+    abstract VarPatternAdmin ownerRole();
+    abstract VarPatternAdmin valueRole();
+    abstract VarPatternAdmin relationOwner();
+    abstract VarPatternAdmin relationValue();
 
-    private final boolean required;
+    abstract boolean required();
 
     /**
      * @throws GraqlQueryException if no label is specified on {@code resourceType}
      */
-    public HasResourceTypeProperty(VarPatternAdmin resourceType, boolean required) {
-        this.resourceType = resourceType;
-        this.required = required;
-
+    public static HasResourceTypeProperty of(VarPatternAdmin resourceType, boolean required) {
         Label resourceLabel = resourceType.getTypeLabel().orElseThrow(GraqlQueryException::noLabelSpecifiedForHas);
 
         VarPattern role = Graql.label(Schema.MetaSchema.ROLE.getLabel());
 
-        VarPattern ownerRole = var().sub(role);
-        VarPattern valueRole = var().sub(role);
+        VarPatternAdmin ownerRole = var().sub(role).admin();
+        VarPatternAdmin valueRole = var().sub(role).admin();
         VarPattern relationType = var().sub(Graql.label(Schema.MetaSchema.RELATION.getLabel()));
 
         // If a key, limit only to the implicit key type
         if(required){
-            ownerRole = ownerRole.label(KEY_OWNER.getLabel(resourceLabel));
-            valueRole = valueRole.label(KEY_VALUE.getLabel(resourceLabel));
+            ownerRole = ownerRole.label(KEY_OWNER.getLabel(resourceLabel)).admin();
+            valueRole = valueRole.label(KEY_VALUE.getLabel(resourceLabel)).admin();
             relationType = relationType.label(KEY.getLabel(resourceLabel));
         }
 
-        this.ownerRole = ownerRole.admin();
-        this.valueRole = valueRole.admin();
-        this.relationOwner = relationType.relates(this.ownerRole).admin();
-        this.relationValue = relationType.admin().getVarName().relates(this.valueRole).admin();
-
-    }
+        VarPatternAdmin relationOwner = relationType.relates(ownerRole).admin();
+        VarPatternAdmin relationValue = relationType.admin().getVarName().relates(valueRole).admin();
 
-    public VarPatternAdmin getResourceType() {
-        return resourceType;
+        return new AutoValue_HasResourceTypeProperty(
```

Comment:
True! I think the issue is mostly just that this class has way more fields than necessary...